python/ray/data/_internal/stream_split_dataset_iterator.py

```diff
@@ -201,3 +236,37 @@ def get(self, output_split_idx: int) -> Optional[ObjectRef[Block]]:
             return block
         except StopIteration:
             return None
+
+    def _barrier(self, split_idx: int) -> int:
+        """Arrive and block until the start of the given epoch."""
+
+        # Decrement and await all clients to arrive here.
+        with self._lock:
+            starting_epoch = self._cur_epoch
+            self._unfinished_clients_in_epoch -= 1
+
+        start_time = time.time()
+        while (
+            self._cur_epoch == starting_epoch and self._unfinished_clients_in_epoch != 0
```

Comment:
Does the read of these two variables need to be protected by reader lock?

Reply:
These should be effectively atomic accesses, so it's safe.